### PR TITLE
tstun, wgengine/magicsock: enable vectorized I/O on Linux.

### PR DESCRIPTION
**Pull request by jwhited**
This commit updates the wireguard-go dependency and implements the
necessary changes to the tun.Device and conn.Bind implementations to
support passing vectors of packets in tailscaled.

Signed-off-by: Jordan Whited <jordan@tailscale.com>
Signed-off-by: James Tucker <james@tailscale.com>
Co-authored-by: James Tucker <james@tailscale.com>

### CODE REVIEW

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -1671,34 +1739,93 @@ func (c *Conn) runDerpWriter(ctx context.Context, dc *derphttp.Client, ch <-chan
...
+}
+
+func (c *Conn) getReceiveBatch() *receiveBatch {
+	batch := c.receiveBatchPool.Get().(*receiveBatch)
+	return batch
+}
+
+func (c *Conn) putReceiveBatch(batch *receiveBatch) {
+	for i := range batch.msgs {
+		batch.msgs[i] = ipv6.Message{Buffers: batch.msgs[i].Buffers}
```
Comment (sailorfrag) [inline wgengine/magicsock/magicsock.go:1753]: Should it set `batch.msgs[i].Buffers[0] = nil`?

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -1671,34 +1739,93 @@ func (c *Conn) runDerpWriter(ctx context.Context, dc *derphttp.Client, ch <-chan
...
+
+		if reportToCaller {
+			return numMsgs, nil
 		}
 	}
 }
 
-// receiveIPv4 receives a UDP IPv4 packet. It is called by wireguard-go.
-func (c *Conn) receiveIPv4(b []byte) (n int, ep conn.Endpoint, err error) {
+func (c *Conn) receiveMultipleIPv4(buffs [][]byte, sizes []int, eps []conn.Endpoint) (int, error) {
```
Comment (sailorfrag) [inline wgengine/magicsock/magicsock.go:1796]: I feel like it wouldn't be overly hard to consolidate this with the ipv6 version, but that's a pre-existing issue and I'm not sure that this is the time to change that.

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -1748,27 +1875,25 @@ func (c *Conn) receiveIP(b []byte, ipp netip.AddrPort, cache *ippEndpointCache,
 	return ep, true
 }
 
-// receiveDERP reads a packet from c.derpRecvCh into b and returns the associated endpoint.
-// It is called by wireguard-go.
-//
-// If the packet was a disco message or the peer endpoint wasn't
-// found, the returned error is errLoopAgain.
-func (c *connBind) receiveDERP(b []byte) (n int, ep conn.Endpoint, err error) {
+func (c *connBind) receiveMultipleDERP(buffs [][]byte, sizes []int, eps []conn.Endpoint) (int, error) {
```
Comment (sailorfrag) [inline wgengine/magicsock/magicsock.go:1878]: this is only receiving 1 packet at a time, is that ok?

**Review by sailorfrag — APPROVED** [review]
I think I understand what the magicsock changes are doing but admittedly it's kind of hard to follow. That's not your fault, that's preexisting. The rest was clear enough.

File: wgengine/magicsock/magicsock.go (hunk 1 above)
Comment (jwhited) [reply wgengine/magicsock/magicsock.go:1753]: This clears all fields in `batch.msgs[i]` when putting back into the pool, but retains the underlying array with capacity 1 in `batch.msgs[i].Buffers[0]`

File: wgengine/magicsock/magicsock.go [hunk 4]
```diff
@@ -1229,6 +1259,44 @@ var errDropDerpPacket = errors.New("too many DERP packets queued; dropping")
 
 var errNoUDP = errors.New("no UDP available on platform")
 
+var (
+	_ ipv6.Message = ipv4.Message{}
```
[the remark this answers is not in the document]
Comment (jwhited) [reply wgengine/magicsock/magicsock.go:1265]: added comment

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (jwhited) [reply wgengine/magicsock/magicsock.go:1880]: I ended up renaming the Receive fns back to their original names. The `multiple` part originated from when we had both single packet and multi packet Receive fns prior to flattening the wg-go conn.Bind API.

There is also more use of the previous names in health/stats/docs that it just made sense to rename back.

File: wgengine/magicsock/magicsock.go (hunk 2 above)
Comment (jwhited) [reply wgengine/magicsock/magicsock.go]: Agreed, we have pre-existing TODOs for this.
